server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/RegexMappingRoutesTest.java

```diff
@@ -0,0 +1,148 @@
+import static org.hamcrest.core.Is.is;
+import static org.mockito.Mockito.mock;
+
+import org.apache.james.core.Domain;
+import org.apache.james.core.User;
+import org.apache.james.dnsservice.api.DNSService;
+import org.apache.james.domainlist.api.DomainList;
+import org.apache.james.domainlist.memory.MemoryDomainList;
+import org.apache.james.rrt.lib.Mapping;
+import org.apache.james.rrt.lib.MappingSource;
+import org.apache.james.rrt.memory.MemoryRecipientRewriteTable;
+import org.apache.james.webadmin.WebAdminServer;
+import org.apache.james.webadmin.WebAdminUtils;
+import org.eclipse.jetty.http.HttpStatus;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import io.restassured.RestAssured;
+import io.restassured.filter.log.LogDetail;
+import io.restassured.http.ContentType;
+
+class RegexMappingRoutesTest {
+
+    private WebAdminServer webAdminServer;
+    private MemoryRecipientRewriteTable memoryRecipientRewriteTable;
+
+    @BeforeEach
+    void beforeEach() throws Exception {
+        DNSService dnsService = mock(DNSService.class);
+        DomainList domainList = new MemoryDomainList(dnsService);
+        memoryRecipientRewriteTable = new MemoryRecipientRewriteTable();
+        memoryRecipientRewriteTable.setDomainList(domainList);
+        domainList.addDomain(Domain.of("domain.tld"));
+
+        webAdminServer = WebAdminUtils
+            .createWebAdminServer(new RegexMappingRoutes(memoryRecipientRewriteTable))
+            .start();
+
+        RestAssured.requestSpecification = WebAdminUtils.buildRequestSpecification(webAdminServer)
+            .setBasePath(RegexMappingRoutes.BASE_PATH)
+            .log(LogDetail.METHOD)
+            .build();
+    }
+
+    @AfterEach
+    void stop() {
+        webAdminServer.destroy();
+    }
+
+    @Test
+    void addRegexMappingShouldReturnNoContentWhenSuccess() {
+        with()
+            .post("james@domain.tld/targets/^[aeiou]:james@domain.tld")
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204)
+            .contentType(ContentType.JSON);
+
+        assertThat(memoryRecipientRewriteTable
+            .getStoredMappings(MappingSource.fromUser(User.fromUsername("james@domain.tld"))))
+            .containsOnly(Mapping.regex("^[aeiou]:james@domain.tld"));
+    }
+
+    @Test
+    void addRegexMappingShouldAllowUserWithoutDomain() {
+        with()
+            .post("jamesdomaintld/targets/^[aeiou]:james@domain.tld")
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204)
+            .contentType(ContentType.JSON);
+
+        assertThat(memoryRecipientRewriteTable
+            .getStoredMappings(MappingSource.fromUser(User.fromUsername("jamesdomaintld"))))
+            .containsOnly(Mapping.regex("^[aeiou]:james@domain.tld"));
+    }
+
+    @Test
+    void addRegexMappingShouldReturnNotFoundWhenSourceIsEmpty() {
+        with()
+            .post("/targets/^[aeiou]:james@domain.tld")
```

Comment:
What happens if the regex contains a `?`

Reply:
If `?` is URL-encoded, code 204 will be returned. Otherwise, code 404 will be return.